Make m_tip_block std::optional

Sjors opened the pull request:
Suggested in https://github.com/bitcoin/bitcoin/pull/31297#discussion_r1844244309

tdb3 reviewed (commented):
Concept ACK

Using `optional` seems much nicer than using a magic value. Planning to circle back to take another look.

l0rinc commented on src/init.cpp:
```diff
@@ -1795,7 +1795,7 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
         WAIT_LOCK(kernel_notifications.m_tip_block_mutex, lock);
         kernel_notifications.m_tip_block_cv.wait(lock, [&]() EXCLUSIVE_LOCKS_REQUIRED(kernel_notifications.m_tip_block_mutex) {
-            return !kernel_notifications.m_tip_block.IsNull() || ShutdownRequested(node);
+            return kernel_notifications.m_tip_block.has_value() || ShutdownRequested(node);
```
I've noticed that we're checking whether the optional has a value in different ways here and in `interfaces.cpp`.
If you change the PR again, please consider unifying:
```suggestion
            return kernel_notifications.m_tip_block || ShutdownRequested(node);
```

Sjors replied:
The reason I did it differently was that `interfaces.cpp` the pattern `blah && blah.value()` makes it clear the first value is a pointer rather than a boolean. That's not clear here. That said, I don't think anyone would be confused either.

l0rinc commented on src/node/interfaces.cpp:
```diff
@@ -973,7 +973,7 @@ class MinerImpl : public Mining
             WAIT_LOCK(notifications().m_tip_block_mutex, lock);
             notifications().m_tip_block_cv.wait_for(lock, timeout, [&]() EXCLUSIVE_LOCKS_REQUIRED(notifications().m_tip_block_mutex) {
-                return (notifications().m_tip_block != current_tip && notifications().m_tip_block != uint256::ZERO) || chainman().m_interrupt;
+                return (notifications().m_tip_block && notifications().m_tip_block.value() != current_tip) || chainman().m_interrupt;
```
Since C++17 we can [safely compare](https://en.cppreference.com/w/cpp/utility/optional/operator_cmp) an `optional<T>` with another `T`:
```C++
template< class T, class U >
constexpr bool operator!=(const optional<T>& opt, const U& value);
(23) (since C++17)
```
If you modify again, consider simplifying to:
```suggestion
                return (notifications().m_tip_block && notifications().m_tip_block != current_tip) || chainman().m_interrupt;
```

TheCharlatan replied:
Can't you skip the first conditional too then?

Sjors replied:
That's probably worth clarifying in a comment...

l0rinc commented on src/test/validation_chainstate_tests.cpp:
```diff
@@ -72,7 +72,8 @@ BOOST_FIXTURE_TEST_CASE(chainstate_update_tip, TestChain100Setup)
     const auto get_notify_tip{[&]() {
         LOCK(m_node.notifications->m_tip_block_mutex);
-        return m_node.notifications->m_tip_block;
+        BOOST_REQUIRE(m_node.notifications->m_tip_block);
```
Does the `BOOST_REQUIRE` serve here as a better error message than what `m_tip_block.value()` would already throw (i.e. `bad_optional_access()`)?

Sjors replied:
I didn't think about it that deeply, but yes.

l0rinc reviewed (commented):
Concept ACK, please see my simplification nits.

sedited reviewed (commented):
lgtm, but I don't think this should be prefixed with kernel: in both the title and the commit message. The point of having the notification interface is that the node code can extend kernel behaviour. The implementation details of that are not part of the kernel.

ryanofsky reviewed (approved):
Code review ACK b283bc8dae0e0b714bdf1828579c67ab94da2cc7.

Can ignore this, but I want to express a slightly dissenting opinion on this change. I don't object to it but wouldn't have recommended it, because checking for `hash.IsNull()` is so widespread in the code, and I think writing interoperable code becomes more error prone when there are multiple ways of describing an unset hash instead of standardizing on one. But on the other hand I understand using `std::optional` can be more self-documenting and intuitive and it can force developers to check for special values (by crashing or triggering undefined behavior) in cases where they might otherwise forget to do that. So this approach probably is a little better in the long term even if it creates more complexity in the short term.

A separate concern is I think b283bc8dae0e0b714bdf1828579c67ab94da2cc7 is using the `optional::value()` method inappropriately. It's good to use the `value()` method in code that wants to catch and throw the `std::bad_optional_access` exception but b283bc8dae0e0b714bdf1828579c67ab94da2cc7 is calling it in cases where exceptions could never be thrown. IMO, unless you actually want this exception, it is better to use `*` and `->` operators because these operators stand out to people used to reading C/C++ code as places where dereferences are happening and don't just look like innocuous method calls. These are the same reasons to prefer the vector `[]` operator over the vector `.at()` method.

Sjors commented:
I dropped the use of `has_value()` https://github.com/bitcoin/bitcoin/pull/31325#discussion_r1851743969 as well as `value()` https://github.com/bitcoin/bitcoin/pull/31325#discussion_r1850680750. Except in the test, where throwing `std::bad_optional_access` is probably useful.

Added a comment: https://github.com/bitcoin/bitcoin/pull/31325/files#r1851747612

Dropped the word "kernel" and added a motivation to the commit message.

l0rinc commented:
ACK 2fff10a656edb8bec7d45433bf431fd233d0fd81

BrandonOdiwuor reviewed (commented):
Concept ACK transitioning `m_tip_block` to `std::optional` which remains unset until a block is connected

ryanofsky commented on src/node/interfaces.cpp:
```diff
@@ -973,7 +973,10 @@ class MinerImpl : public Mining
             WAIT_LOCK(notifications().m_tip_block_mutex, lock);
             notifications().m_tip_block_cv.wait_for(lock, timeout, [&]() EXCLUSIVE_LOCKS_REQUIRED(notifications().m_tip_block_mutex) {
-                return (notifications().m_tip_block != current_tip && notifications().m_tip_block != uint256::ZERO) || chainman().m_interrupt;
+                // Although C++17 allows safe comparison of optional<T> with
+                // another T, we need to wait for m_tip_block to be set AND
```
In commit "Make m_tip_block an std::optional" (2fff10a656edb8bec7d45433bf431fd233d0fd81)

IMO would be less confusing to drop comment about c++17 and just say "We need to wait for m_tip_block to be set AND for the value to be different than the current_tip value" which explains why there are two conditions.

ryanofsky reviewed (commented):
Code review ACK 2fff10a656edb8bec7d45433bf431fd233d0fd81

Sjors commented:
Rebased (just in case) and shortened the comment: https://github.com/bitcoin/bitcoin/pull/31325#discussion_r1870138808

l0rinc commented:
ACK 5c48d098dc300ae47f2ffeb9305f58bd6352152d
The only change was to the comment which I find less technical and more descriptive 👍

l0rinc commented on src/node/interfaces.cpp:
```diff
@@ -973,7 +973,9 @@ class MinerImpl : public Mining
             WAIT_LOCK(notifications().m_tip_block_mutex, lock);
             notifications().m_tip_block_cv.wait_for(lock, timeout, [&]() EXCLUSIVE_LOCKS_REQUIRED(notifications().m_tip_block_mutex) {
-                return (notifications().m_tip_block != current_tip && notifications().m_tip_block != uint256::ZERO) || chainman().m_interrupt;
+                // We need to wait for m_tip_block to be set AND for the value
+                // to be different than the current_tip value.
```
nit: https://languagetool.org complains here
```
Did you mean 'different from'? 'Different than' is often considered colloquial style. 
Incorrect:  Are human beings any different than animals? 
Correct:  Are human beings any different from animals? 
```

```suggestion
                // We need to wait for m_tip_block to be set AND for the value
                // to differ from the current_tip value.
```

Sjors replied:
Taken

l0rinc commented:
ACK fe6487ca6ab005faab9f1d565740f2259c04cd16

ryanofsky reviewed (approved):
Code review ACK fe6487ca6ab005faab9f1d565740f2259c04cd16 just taking suggesting code comment changes since last review. (Thanks for the updates!)

luke-jr commented:
Not sure this is an improvement. Seems like a case of "two ways to be omitted" which has caused issues in the past.

l0rinc commented on src/init.cpp:
```diff
@@ -1797,7 +1797,7 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
         WAIT_LOCK(kernel_notifications.m_tip_block_mutex, lock);
         kernel_notifications.m_tip_block_cv.wait(lock, [&]() EXCLUSIVE_LOCKS_REQUIRED(kernel_notifications.m_tip_block_mutex) {
-            return !kernel_notifications.m_tip_block.IsNull() || ShutdownRequested(node);
+            return kernel_notifications.TipBlock() || ShutdownRequested(node);
```
👍 for unifying

l0rinc commented on src/node/kernel_notifications.cpp:
```diff
@@ -52,6 +52,7 @@ kernel::InterruptResult KernelNotifications::blockTip(SynchronizationState state
 {
     {
         LOCK(m_tip_block_mutex);
+        Assert(index.GetBlockHash() != uint256{});
         m_tip_block = index.GetBlockHash();
```
I think this should either be an `assert` or an `Assume` instead.
> Assume
>  * Should be used to run non-fatal checks. In debug builds it behaves like
> *   Assert()/assert() to notify developers and testers about non-fatal errors.
>  *   In production it doesn't warn or log anything.

And in most places I saw the `!....IsNull()` pattern instead:
https://github.com/bitcoin/bitcoin/blob/master/src/txdb.cpp#L97, https://github.com/bitcoin/bitcoin/blob/master/src/wallet/scriptpubkeyman.cpp#L1000, https://github.com/bitcoin/bitcoin/blob/master/src/primitives/transaction.cpp#L77, https://github.com/bitcoin/bitcoin/blob/master/src/validation.cpp#L2490, etc have have

And it would be simpler to check after assignment.

```suggestion
        m_tip_block = index.GetBlockHash();
        assert(!m_tip_block.IsNull());
```

Sjors replied:
I think newer code is trying to avoid `assert` to prevent opening production nodes open to attacks like https://bitcoincore.org/en/2024/10/08/disclose-blocktxn-crash/

Looking at `developer-notes.md` section on asserts, I think `Assume` is better. However it depends on whether a mistake here is dangerous, which hard to say.

I find `IsNull()` too ambiguous for use with an `std::optional`.

l0rinc replied:
> I find IsNull() too ambiguous for use with an std::optional.

My mistake, I meant to write `!m_tip_block->IsNull()`, i.e. it's not on the optional

l0rinc replied:
Or you only want to check that it's either empty or if not it's not 0? In that case this seems like the simplest indeed

Sjors replied:
`GetBlockHash()` is not an `std::optional`, so I only want to make sure it's not `0`.

After this is checked, we assign the value to `m_tip_block` which _is_ an `std::optional`, but at that point I'm not worried that `m_tip_block` remains unset.

l0rinc commented on src/node/kernel_notifications.cpp:
```diff
@@ -99,6 +100,14 @@ void KernelNotifications::fatalError(const bilingual_str& message)
+std::optional<uint256> KernelNotifications::TipBlock()
+{
+    AssertLockHeld(m_tip_block_mutex);
+    Assert(!m_tip_block || m_tip_block.value() != uint256{});
```
we're already doing validation during assignment, this seems redundant to me

Sjors replied:
It is now, but maybe we introduce assignment somewhere else and forget to check it there. This would catch such a mistake.

It's a bit overkill, which we could drop once the whole codebase uses a single pattern.

l0rinc replied:
> It's a bit overkill, which we could drop once the whole codebase uses a single pattern.

Valid argument

l0rinc commented on src/node/kernel_notifications.h:
```diff
@@ -56,15 +56,19 @@ class KernelNotifications : public kernel::Notifications
     std::atomic<int>& m_exit_status;
     node::Warnings& m_warnings;
+
+    // May not be ZERO.
```
Same, comment seems excessive, the code already makes this clear

l0rinc commented on src/test/validation_chainstate_tests.cpp:
```diff
@@ -72,7 +72,8 @@ BOOST_FIXTURE_TEST_CASE(chainstate_update_tip, TestChain100Setup)
         LOCK(m_node.notifications->m_tip_block_mutex);
-        return m_node.notifications->m_tip_block;
+        BOOST_REQUIRE(m_node.notifications->TipBlock());
+        return m_node.notifications->TipBlock().value();
```
nit:
```suggestion
        return *m_node.notifications->TipBlock();
```

Sjors replied:
Done

Sjors commented:
I switched from `Assert` to `Assume`.

l0rinc commented:
ACK fc0b5f3b77166aad9fb288b8eb9185bb2dc7496c

Sjors commented:
Rebased after #31346.

l0rinc commented:
ACK 43f23cd29de13ea600bbe1172cf1ac0d535f66ee

Sjors commented:
@maflcko thoughts, since you suggested it?

fjahr commented on src/node/kernel_notifications.cpp:
```diff
@@ -99,6 +100,14 @@ void KernelNotifications::fatalError(const bilingual_str& message)
+std::optional<uint256> KernelNotifications::TipBlock()
+{
+    AssertLockHeld(m_tip_block_mutex);
+    Assume(!m_tip_block || m_tip_block.value() != uint256{});
```
nit: I tend to agree with other commenters that explicitly checking for non-zero and the comment about it might not be needed but I am not too passionate about it. Similarly I would add to that that this line here could have used `uint256::ZERO` like the code that was changed previously.

Sjors replied:
There was another pull request where someone suggested the use of `uint256{}` instead of `uint256::ZERO`, though I forgot where. The codebase uses both. Though in this context perhaps `ZERO` makes the intention more clear.

fjahr reviewed (commented):
utACK 43f23cd29de13ea600bbe1172cf1ac0d535f66ee

Sjors commented on src/node/kernel_notifications.h:
```diff
@@ -56,10 +56,11 @@ class KernelNotifications : public kernel::Notifications
 
     Mutex m_tip_block_mutex;
     std::condition_variable m_tip_block_cv GUARDED_BY(m_tip_block_mutex);
+
```
I find that it makes the next comment more readable.

Sjors replied:
I have to retouch, so will drop it.

tdb3 commented on src/node/kernel_notifications.cpp:
```diff
@@ -99,6 +100,14 @@ void KernelNotifications::fatalError(const bilingual_str& message)
+std::optional<uint256> KernelNotifications::TipBlock()
+{
+    AssertLockHeld(m_tip_block_mutex);
+    Assume(!m_tip_block || m_tip_block.value() != uint256{});
```
nit:
Using `uint256::ZERO` seems clearer to me, but I don't feel too strongly about it.

Sjors replied:
I droppend this check, only keeping the setter.

tdb3 commented on src/node/kernel_notifications.cpp:
```diff
@@ -52,6 +52,7 @@ kernel::InterruptResult KernelNotifications::blockTip(SynchronizationState state
 {
     {
         LOCK(m_tip_block_mutex);
+        Assume(index.GetBlockHash() != uint256{});
```
nit:
Same here (`uint256::ZERO`)

Sjors replied:
Done here

tdb3 reviewed (approved):
Code review ACK 43f23cd29de13ea600bbe1172cf1ac0d535f66ee

Left a pico nit, but nothing blocking.

Sjors committed:
Ensure m_tip_block is never ZERO

To avoid future code changes from reintroducing the ambiguity fixed
by the previous commit, mark m_tip_block private and Assume that
it's not set to uint256::ZERO.

Sjors commented:
I replaced the use of `uint256{}` with `uint256::ZERO`, since it makes the intention more clear. I also limited the not-ZERO check to just the setter and dropped the comment.

Also dropped the whitespace change in `kernel_notifications.h` in the first commit.

tdb3 commented on src/node/kernel_notifications.h:
```diff
@@ -59,12 +59,14 @@ class KernelNotifications : public kernel::Notifications
     std::atomic<int>& m_exit_status;
     node::Warnings& m_warnings;
+
+    std::optional<uint256> m_tip_block GUARDED_BY(m_tip_block_mutex){};
```
I could be mistaken, but the `{}` initializer here might be initializing to uint256::ZERO instead of `nullopt`. Intended?

Sjors replied:
I'll drop the `{}` so it should be clear that it inits to `nullopt`. I'm fairly sure the current syntax also does that and is not the same as `{uint256{}}`.

tdb3 reviewed (commented):
Approach ACK

Sjors committed:
Make m_tip_block an std::optional

This change avoids ambiguity when no tip is connected and it is
compared to uint256::ZERO.

tdb3 reviewed (approved):
code review re ACK 81cea5d4ee0e226f7c7145072de85829f4166ec9

l0rinc reviewed (approved):
ACK 81cea5d4ee0e226f7c7145072de85829f4166ec9

fjahr commented:
re-ACK 81cea5d4ee0e226f7c7145072de85829f4166ec9
